src/stores/utils/contract.js

```diff
@@ -66,16 +69,42 @@ export const totalBurntCoins = async (contract) => {
   return new BN(burntCoins)
 }
 
-export const getBridgeValidators = async (bridgeValidatorContract) => {
-  let ValidatorAdded = await bridgeValidatorContract.getPastEvents('ValidatorAdded', {fromBlock: 0});
-  let ValidatorRemoved = await bridgeValidatorContract.getPastEvents('ValidatorRemoved', {fromBlock: 0});
-  let addedValidators = ValidatorAdded.map(val => {
-    return val.returnValues.validator
-  })
-  const removedValidators = ValidatorRemoved.map(val => {
-    return val.returnValues.validator
+export const getValidatorList = async (address, eth) => {
+  const v1Contract = new eth.Contract(BridgeValidatorsV1Abi, address);
+  const v1ValidatorsEvents = await getValidatorEvents(v1Contract)
+
+  const contract = new eth.Contract(BRIDGE_VALIDATORS_ABI, address);
+  const deployedAtBlock = await getDeployedAtBlock(contract);
+  const validatorsEvents = await getValidatorEvents(contract, deployedAtBlock)
+
+  const rewardableContract = new eth.Contract(REWARDABLE_BRIDGE_VALIDATORS_ABI, address);
```

Comment:
Does it mean that we need to use the ABI of the rewardable validator contract even if the fee manager is not in action?

Another point is: if we use the rewardable validators contract the method `validatorList` can be invoked to get the list of actual validators.

Reply:
See this comment on why we use different ABIs https://github.com/poanetwork/bridge-ui/pull/184#discussion_r269309818

> Another point is: if we use the rewardable validators contract the method validatorList can be invoked to get the list of actual validators.

Thanks for the suggestion. I think there is room for improvement here. We can try to call `validatorList` to get the list as first intent. If it doesn't fail and return the list, there are no other required action. If it fails then we can continue to get the list of validator by listening to events from contracts using ABIs from v1 and v2. What do you think?

Reply:
Could you describe a case when `the list of validator by listening to events from contracts using ABIs from v1 and v2` is applicable with the new UI?

Reply:
For example if we take a bridge that was deployed using v1 of contracts as POA-POA20 bridge, there are several v1 `ValidatorAdded` events. Then if for some reason contracts are upgraded to v2, adding or removing validators will generate v2 `ValidatorAdded`/`ValidatorRemoved` events. So we need to listen to all of that events in order to generate and display the correct list of validators. With the changes introduced mentioned here https://github.com/poanetwork/bridge-ui/pull/184#issuecomment-481661210 we don't need anymore to make one request with v1 ABI and other with v2 ABI to get all events. Only one request will be performed and the events will be parse manually without using any of the ABIs, only the topics hashes to detect the events.

Reply:
Right. But I assume that after upgrade the contracts will keep information about validators in the list so we don't need to listen events anymore at all in order to discover actual list of the bridge authorities.

My suggestion is to introduce the functionality to keep the list of validators in the BridgeValidators contract, so after the contracts upgrade from v1 to v2 we will don't need to track `ValidatorAdded`/`ValidatorRemoved` events.

Reply:
PR on contracts side was created to add `validatorList` method to `BridgeValidators` contract: https://github.com/poanetwork/poa-bridge-contracts/pull/167#event-2269809275 

Also updated the logic here https://github.com/poanetwork/bridge-ui/pull/184/commits/000864c5670dac76d9e449c38c686b5701b76f60

To get the list of validators now we call `validatorList` from validators contract. If that call fails (as in the case of xDai bridge that validators implementation doesn't have that method) or return an empty array (as in the case of ETC like chains. - tested with ETC-WETC bridge, on ETC side it returns empty array) then the logic to get validators using events is applied

Reply:
Ouch! Completely forgot that the contracts for ETC was compiled with the old version of Solidity that's why we cannot return dynamic arrays from the methods there. Code features in the project trend to go our of control.

Reply:
I will look at the PR in the bridge contracts repo.